### PR TITLE
Add a warning when the host volume in the compose file is masked by a data volume

### LINKED ISSUE #2308: Container volume masks host volume in the compose file
I have a little snippet in `docker-compose.yml` like:

```
volumes:
  - ${HOST_PATH}:/container/path
```
If I change the value of `HOST_PATH` any subsequent `up` or `build`, the container will still mount the volume with the *old* value of `HOST_PATH`.

Is this because `volumes` is not a run time option like `docker -v` or is this a bug?

### PR DESCRIPTION
**Pull request by dnephin**
Fixes #2308

Context in the discussion below

### CODE REVIEW

**Comment by shin-** [comment]
Is this warning going to show even if the host path doesn't change? I can see users being confused by that when there's nothing wrong with their service config in the first place.

**Comment by dnephin** [comment]
It will show anytime the container is ignoring the `volume:` setting in the Compose file.  I can't think of a reason that you'd define a host volume mapping in the config, but for some reason want to ignore it and use the data volume from a container.

The most common case for this is when there is a `VOLUME` in their Dockerfile for a service.

If the user starts with a simple Compose file (one without volumes) and runs `docker-compose up` they will get a container with a container data volume. They later decide to add a host mapping for that service at the same path. When they run `docker-compose up` again, they'll see this warning, because the config change they made is being ignored.

If they `docker-compose rm` before running `up`, there will be no warning.

If they never `up` before adding the `volume: ` section to the service in the Compose file, there will be no warning.

I think that is how it should  be working now. It should only warn when something is actually being ignored from the Compose file.

**Comment by dnephin** [comment]
Looking at the test again, I see what you're saying. I need to look at it again to see if it's actually testing the right thing.

**Comment by dnephin** [comment]
I fixed up the logic and the test. It was warning anytime a host volume existed at the same path as an image volume, even if the host volume was being used.  

Now it only warns when the host volume is being ignored.

File: compose/service.py [hunk 1]
```diff
@@ -951,9 +952,26 @@ def get_container_data_volumes(container, volumes_option):
...
+    container_volumes = dict(
+        (volume.internal, volume.external)
+        for volume in container_volumes)
+
+    for volume in volumes_option:
+        if container_volumes.get(volume.internal) != volume.external:
+            log.warn((
+                "Service \"{service}\" is using volume \"{volume}\" from the "
+                "previous container. Host mapping \"{host_path}\" has no effect. "
+                "Remove the existing containers to use the host volume mapping."
```
Comment (aanand) [inline compose/service.py]: Could print the command to use here - `docker-compose rm {service}`

Comment (dnephin) [inline compose/service.py]: Ya, I started with that, but it gets a bit messy if they've used a different name for the command. I would use `sys.argv[0]`, but that often has a full path on it, so I'd have to `dirname()` to truncate.

I could make it `"%s rm %s" % (dirname(sys.argv[0]), service)`

Comment (aanand) [inline compose/service.py]: True, but I don't think we really need to serve that case. We already print `docker-compose` in several other places.

Comment (dnephin) [inline compose/service.py]: ok

**Comment by dnephin** [comment]
`docker-compose rm {service}` added to the warning message

**Comment by aanand** [comment]
LGTM

**Comment by doozr** [comment]
I'm seeing a problem with this implementation; if I start a container with a host volume mounted then change the configuration in some other way (e.g. adding an evironment variable) then I still get the warning when the container is recreated.

Furthermore, if I add a host volume that was not previously there, either as a host or data volume, then the warning appears.

In summary:

* if a container is recreated with an identical host volume, it still shows the warning even though the correct directory is mounted
* if a container is recreated with a new host volume that never previously existed as either host nor data volume, the warning appears and makes no sense.

**Comment by doozr** [comment]
For example, take this original file:

    test:
        image: busybox
        volumes:
            - "/home/vagrant:/vagrant"
        environment:
            TEST: 1

If I modify it thus and run `docker-compose up -d`, I would not expect the warning, but the warning still appears:

    test:
        image: busybox
        volumes:
            - "/home/vagrant:/vagrant"
        environment:
            TEST: 2

Then, if I add a new volume that has never been declared:

    test:
        image: busybox
        volumes:
            - "/home/vagrant:/vagrant"
            - "/tmp:/hosttmp"
        environment:
            TEST: 2

Then I get a warning about both volumes.

**Comment by dnephin** [comment]
Thanks for the bug report. I've got a fix out here: #2390

**Pull request merged**